### ISSUE TITLE
Unconstrained arrays **'d in headers are still being allocated via Trick in input file

### ISSUE THREAD

**Issue by excaliburtb** [nasa/trick#518]
An unconstrained array is set in a header file as checkpoint I/O '**' which prevents it from being checkpoint/restored. However, when being assigned to via the input file, the I/O attributes are not checked before being allocated as a Trick-managed variable and assigned to the variable.

**Comment by iamthad** [comment]
What do you mean by unconstrained array? Do you mean a pointer? Can you provide an example of the declaration you're describing?

**Comment by excaliburtb** [comment]
Yes, "unconstrained array" is the nomenclature used to differentiate a pointer that is simply pointing to another instance of something versus a pointer to an array of data that is intended to be allocated in the input file or in an early initialization routine somewhere.

class Foo {
public:
char * name; /* trick_chkpt_io(**) */
double * vals; /* trick_chkpt_io(**) */
};
members like that.

**Comment by iamthad** [comment]
My understanding is that checkpoint IO is (now?) distinct from other types of IO, including the variable server. `trick_chkpnt_io` handles the first, while `trick_io` handles the second. 

I'm not sure that either flag will affect the behavior of the input file, since that's all handled by the SWIG-generated code.

**Comment by dbankieris** [comment]
That's correct. Python access is controlled only by privacy. If it's public, you can set it. If it's not, you can't.

**Comment by iamthad** [comment]
Perhaps you can clarify what the behavior you expect is, @excaliburtb.

**Comment by excaliburtb** [comment]
class Foo {
char * name;
};

input file:
foo.name = "Hello!"

Trick is under-the-hood allocating name for the user using TMM_strdup(). It's nice of Trick to do so, but it is a problem if the "name" variable is checkpoint **'d which means that the allocation needs to be done with new or malloc by the user.

**Comment by dbankieris** [comment]
@excaliburtb I don't know if this was different 18 months ago. An anonymous string is indeed checkpointed, but the `**`'d out variable is not. 

```c++
// S_define
#include "sim_objects/default_trick_sys.sm"
class Sandbox : public Trick::SimObject {
    public:
    char* potato; // trick_chkpt_io(**)
    Sandbox() {}
};
Sandbox sandbox;
```

```python
# input.py
sandbox.potato = "Potato"
sandbox.potato = "Potato2"
sandbox.potato = "Potato3"
trick.checkpoint_pre_init(True)
```

```c++
// chkpnt_pre_init
...
trick_anon_local_3 = "Potato";
trick_anon_local_4 = "Potato2";
trick_anon_local_5 = "Potato3";
// no sandbox.potato assignment
...
```
This results in some wasted memory when the checkpoint is reloaded. Is there some other problem I'm missing?

**Comment by excaliburtb** [comment]
Yes, the problem is that the original allocation is deleted. and the new allocation is never assigned back to the pointer. So the potato variable is pointing to freed memory post checkpoint restart.

**Comment by dbankieris** [comment]
Ah, I see. The allocation is made in [`typemap_in_1dp`](https://github.com/nasa/trick/blob/4f1ba26411e3ba2da914730a93cb095d10b46927/include/trick/swig/swig_int_templates.hh#L193), which is called from the `_wrap_*_set` function of the SWIG-generated `*_py.cpp` file. That code somehow follows from [`swig_int_typemap.i`](https://github.com/nasa/trick/blob/4f1ba26411e3ba2da914730a93cb095d10b46927/include/trick/swig/swig_int_typemap.i#L76), and now I'm in way over my head. It looks like we'd have to determine whether the allocation should be Trick-managed or external and pass that information all the way down to `typemap_in_1dp`. It's up to @alexlin0 from here!

**Comment by excaliburtb** [comment]
Right. IO `**` implies that Trick doesn't manage it at all, so it probably needs a strdup() call or the like instead.
